### PR TITLE
InceptionTime Model for Time Series

### PR DESCRIPTION
**Pull request by codeboy5**
This PR will contain the implementation of InceptionTime Model and it's use for classification and regression task.

Some of the commits from the PR #253 are also in this PR, but will take care of them when that PR is merged.

### CODE REVIEW

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    out2 = m.maxconvpool(input)
+    x = cat(out1..., out2, dims = 2)
+    return Flux.relu(m.bn(x))
+end
+
+struct InceptionBlock
+    depth::Int
+    inception::Any
+    shortcut::Any
+end
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:50]: I think all this could be done using Flux's built-in branching/sequencing layers. `Chain` and `Parallel` in particular.

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+Flux.@functor InceptionModule
+
+# Model Output
+function (m::InceptionModule)(x)
+    input = x
+    x = m.bottleneck(x)
+    out1 = [m.convs[i](x) for i in range(1,stop=3)]
+    out2 = m.maxconvpool(input)
+    x = cat(out1..., out2, dims = 2)
+    return Flux.relu(m.bn(x))
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:43]: Note that `BatchNorm` takes an activation function. I think it's faster than applying one externally, but either way it's more correct.

File: FastTimeSeries/src/models.jl [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+blockmodel(inblock::TimeSeriesRow, outblock::OneHotTensor{0}, backbone)
+
+Construct a model for time-series classification.
+"""
+function blockmodel(inblock::TimeSeriesRow,
+                outblock::OneHotTensor{0}, 
+                backbone)
+    data   = rand(Float32, inblock.nfeatures, 32, inblock.obslength)
+    # data   = [rand(Float32, inblock.nfeatures, 32) for _ ∈ 1:inblock.obslength]
+    output = backbone(data)
```
Comment (ToucheSir) [inline FastTimeSeries/src/models.jl:12]: [`Flux.outputsize`](https://fluxml.ai/Flux.jl/stable/utilities/#Flux.outputsize) was designed for precisely this.

File: FastTimeSeries/src/models/InceptionTime.jl (hunk 1 above)
Comment (codeboy5) [reply FastTimeSeries/src/models/InceptionTime.jl:50]: Yeah just looked up the documentation, I think using Parallel for this would work. Would make the changes.

**Comment by codeboy5** [comment]
Have to fix a couple of issues with the residual part.

**Comment by codeboy5** [comment]
Hey @ToucheSir @darsnack @lorenzoh,
So If i do use the residual layers, I seem to be getting a size mismatch during the backprop. The forward pass and loss computation works fine. I am still trying to figure out where exactly the issue is. If you see any issues by just taking a quick look at the code, it would help. Thanks

**Comment by ToucheSir** [comment]
Try building up your model bit-by-bit and checking for size mismatches on each submodule with a dummy input. `Flux.outputsize` can also be very useful here.

**Comment by codeboy5** [comment]
> Try building up your model bit-by-bit and checking for size mismatches on each submodule with a dummy input. `Flux.outputsize` can also be very useful here.

Yeah I tried doing that. I dont understand how the forward pass works fine but the backprop gives an error. Will figure it out.

**Comment by ToucheSir** [comment]
Since only `InceptionModule` has been ported over so far, you could save some time by just testing with that. Feel free to drop a MWE and stacktrace here if you get stuck.

`InceptionBlock` will need to receive the same treatment at some point. One resource which might help are the 2D [Inception image models](https://github.com/FluxML/Metalhead.jl/tree/master/src/convnets/inception)  in Metalhead. Those make extensive use of Flux container layers.

**Comment by codeboy5** [comment]
The code can be tested with the following snipet :- 

```Julia
data, blocks = load(datarecipes()["natops"]);
task = TSClassificationSingle(blocks, data);
traindl, validdl = taskdataloaders(data, task, 16);
callbacks = [ToGPU(), Metrics(accuracy)];
model = FastTimeSeries.Models.InceptionTime(24, 6);
learner = Learner(model, tasklossfn(task); data=(traindl, validdl), optimizer=ADAM(), callbacks = [ToGPU(), Metrics(accuracy)]);
fitonecycle!(learner, 10, 0.01)
```

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        else
+            push!(blocks, Chain(blk...))
+        end
+    end
+    return Chain(blocks...)
+
+end
+
+function changedims(X)
+    X = permutedims(X, (2, 1, 3))
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:69]: I thought permuting was not required for the conv-based models? Or are .ts datasets read in as `features x timesteps x batch`?

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+"""
+function InceptionModule(ni::Int, nf::Int, ks::Int = 40, bottleneck::Bool = true)
+    ks = [ks ÷ (2^i) for i in range(0, stop = 2)]
+    ks = [ks[i] % 2 == 0 ? ks[i] - 1 : ks[i] for i in range(1, stop = 3)]  # ensure odd ks
+    bottleneck = ni > 1 ? bottleneck : false
+
+    bottleneck_block = bottleneck ? Conv1d(ni, nf, 1, bias = false) : identity
+
+    convs_layers =
+        [Conv1d(bottleneck ? nf : ni, nf, ks[i], bias = false) for i in range(1, stop = 3)]
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:14]: I think it makes more sense to have `Conv1d` be lowercase `conv1d`, but that's a minor thing.

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 6]
```diff
@@ -0,0 +1,83 @@
+"""
+    InceptionModule(ni::Int, nf::Int, ks::Int = 40, bottleneck::Bool = true)
+
+TBW
+"""
+function InceptionModule(ni::Int, nf::Int, ks::Int = 40, bottleneck::Bool = true)
+    ks = [ks ÷ (2^i) for i in range(0, stop = 2)]
+    ks = [ks[i] % 2 == 0 ? ks[i] - 1 : ks[i] for i in range(1, stop = 3)]  # ensure odd ks
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:8]:
```suggestion
    ks = [iseven(ks[i]) ? ks[i] - 1 : ks[i] for i in 1:3]  # ensure odd ks
```
Almost always prefer range syntax to the function (I haven't added suggestions for above and below, but this applies there too). I think it may also help not to use the same variable name (`ks`) for the original input parameter and your subsequent list of kernel sizes.

**Review by ToucheSir — COMMENTED** [review]
Just a couple comments on the WIP here. Mostly looks good :)

File: FastTimeSeries/src/models/InceptionTime.jl (hunk 4 above)
Comment (codeboy5) [reply FastTimeSeries/src/models/InceptionTime.jl]: When we load the data, it has the shape `num_samples  x features x timesteps`. Intuitively this look easier to understand.

**Comment by codeboy5** [comment]
Hey guys, so the model is working for regression task too now, will push the notebook in a while.

**Comment by ToucheSir** [comment]
I have some suggestions for the notebook, but that can wait for a follow-up PR. Will try to have a look at the model again in the next couple of days.

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        end
+        if residual && d ÷ 3 <= length(shortcut)
+            skp = shortcut[d÷3]
+            push!(blocks, Parallel(+, Chain(blk...), skp))
+        else
+            push!(blocks, Chain(blk...))
+        end
+    end
+    return Chain(blocks...)
+
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:65]:
```suggestion
```

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+"""
+    InceptionBlock(ni::Int, nf::Int = 32, residual::Bool = true, depth::Int = 6)
+
+TBW
+"""
+function InceptionBlock(ni::Int, nf::Int = 32, residual::Bool = true, depth::Int = 6)
+    inception = []
+    shortcut = []
+
+    for d in range(1, stop = depth)
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:32]: Still many wild `range`s floating around ;)

File: FastTimeSeries/src/models/RNN.jl [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+- `outsize`: The output size of the final classifier block. For single classification tasks,
+    this would be the number of classes.
+- `recout`: The output size of the `recbackbone` architecture.
+- `dropout_rate`: Dropout probability for the dropout layer.
+"""
+
+function RNNModel(recbackbone;
+                outsize,
+                recout,
+                kwargs...)
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/RNN.jl:25]:
```suggestion
function RNNModel(recbackbone; outsize, recout)
```
Extra `kwargs` don't appear to be used here?

File: FastTimeSeries/src/models/InceptionTime.jl [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+end
+
+changedims(X) = permutedims(X, (2, 1, 3))
+
+"""
+    InceptionTime(c_in::Int, c_out::Int, seq_len = nothing, nf::Int = 32)
+
+TBW
+"""
+function InceptionTime(c_in::Int, c_out::Int, seq_len = nothing, nf::Int = 32)
```
Comment (ToucheSir) [inline FastTimeSeries/src/models/InceptionTime.jl:75]: These `in` and `out` arg names should be harmonized across the various time series models (i.e. InceptionTime and StackedLSTM).

File: FastTimeSeries/src/models/InceptionTime.jl (hunk 8 above)
Comment (codeboy5) [reply FastTimeSeries/src/models/InceptionTime.jl:32]: Yeah, will fix that in the next commit.

**Review by ToucheSir — COMMENTED** [review]
Mostly LGTM once docstrings are added and commented code is restored/removed.

**Review by ToucheSir — APPROVED** [review]
Sorry about the delay on reviewing this. Let's get it merged.
